README.md
DataCombine
======

Christopher Gandrud

Version 0.2.23
[![Build Status](https://travis-ci.org/christophergandrud/DataCombine.png)](https://travis-ci.org/christophergandrud/DataCombine) [![CRAN Version](http://www.r-pkg.org/badges/version/DataCombine)](https://cran.r-project.org/package=DataCombine) ![CRAN Downloads](http://cranlogs.r-pkg.org/badges/last-month/DataCombine) ![CRAN Total Downloads](http://cranlogs.r-pkg.org/badges/grand-total/DataCombine)


Please report any **bugs** or **suggestions** at:
<https://github.com/christophergandrud/DataCombine/issues>.

## Motivation and Functions

**DataCombine** is a set of miscellaneous tools intended to make combining
data sets--especially time-series cross-section data--easier. The package is
continually being developed as I turn lines of code that I frequently use into
single functions. It currently includes the following functions:

- `CasesTable` function added to report cases after list-wise deletion of
missing values for time-series cross-sectional data.

- `change`: calculates the absolute, percentage, and proportion change from
a specified lag, including within groups.

- `CountSpell`: function that returns a variable counting the spell number
for an observation. Works with grouped data.

- `dMerge`: merges 2 data frames and report/drop/keeps only duplicates.

- `DropNA`: drops rows from a data frame when they have missing (`NA`) values on a
given variable(s).

- `FillDown`: fills in missing (`NA`) values with the previous non-missing value

- `FillIn`: fills in missing values of a variable from one data frame with the
values from another variable.

- `FindDups`: find duplicated values in a data frame and subset it to either
include or not include them.

- `FindReplace`: replaces multiple patterns found in a character string column
of a data frame.

- `grepl.sub`: subsets a data frame if a specified pattern is found in a
character string.

- `InsertRow`: allows user to insert a row into a data frame. Largely
implements: [Ari B. Friedman's function](http://stackoverflow.com/a/11562428).

- `MoveFront`: moves variables to the front of a data frame. This can be useful
if you have a data frame with many variables and want to move a variable or
variables to the front.

- `NaVar`: create new variable(s) indicating if there are missing values in
other variable(s).

- `shift`: creates lag and lead variables, including for time-series
cross-sectional data. The shifted variable is returned to a new vector. This
function is largely based on
[TszKin Julian's shift function](http://ctszkin.com/2012/03/11/generating-a-laglead-variables/).

- `slide`: creates lag and lead variables, including for time-series
cross-sectional data. The slid variable are added to the original data frame.
This expands the capabilities of `shift`.

- `slideMA`: creates a moving average for a period before or after each time
point for a given variable.

- `SpreadDummy`: spread a dummy variable (1's and 0') over a specified time
period and for specified groups.

- `StartEnd`: finds the starting and ending time points of a spell, including
for time-series cross-sectional data.

- `rmExcept`: removes all objects from a workspace except those specified by the
user.

- `TimeExpand`: expands a data set so that it includes an observation for each
time point in a sequence. Works with grouped data.

- `TimeFill`: creates a continuous `Unit`-`Time`-`Dummy` data frame from a data
frame with `Unit`-`Start`-`End` times.

- `VarDrop`: drops one or more variables from a data frame.

## Updates

I will continue to add to the package as I build data sets and run across other
pesky tasks I do repeatedly that would be simpler if they were completed by
a single function.

## Installation

**DataCombine** is on [CRAN](https://cran.r-project.org/).

You can also install the most recent stable version with `install_github` from
the [devtools](https://github.com/hadley/devtools):

```{r}
devtools::install_github('christophergandrud/DataCombine')
```


NEWS
## Version 0.2.23

- `DropNA` should now be compatible with `tbl_df` data frames.

- Introduced (a largely internal function) `pull` for extracting single columns
from data frames (including `tbl_df`) and returning them as vectors.
The function is based on Tommy O'Dell's:
<http://stackoverflow.com/a/24730843/1705044>.

## Version 0.2.22

- Fixed a bug in `CasesTable` where multiple variables would cause an error.

- Minor example update and internal code formatting improvements to
`TimeFill`.

## Version 0.2.21

- `CasesTable` function added to report cases after listwise deletion of
missing values for time-series cross-sectional data.

- Fixed a bug where `CountSpell` would fail if all `SpellVar` values were
the same.

- Improved documentation.

## Version 0.2.20

- Internal changes to enable compatability with dplyr version 0.4.4.

- The `Var` argument in `dMerge` is deprecated. Use `by` instead to match
`merge` syntax more closely.

- `DropNA` drops all NAs from the supplied data frame if `Var` is missing.

## Version 0.2.19

- Fixed a bug in `dMerge` that overwrote users `suffixes` specification.
Thanks to @hplieninger for spotting this.

- `FillIn` intelligently fails if `Var1` and `Var2` do not exist in their
respective data.frames.

## Version 0.2.18

- `SpreadDummy` now starts by coercing the `data` to a data frame, so that it
works more smoothly with dplyr data tables.

## Version 0.2.17

- `change` calculate the changes (absolute, percent, and proportion) changes
from a specified lag, including within groups. Replaced `PercChange` which
has the same syntax, but did not find absolute changes.

## Version 0.2.16

- `FindDups` code and documentation improvements.

## Version 0.2.15

- `FindDups` for finding duplicated values in a data frame and subsetting it
to either include or not include them.

## Version 0.2.14

- `grepl.sub` now uses dot notation to pass arguments to `grepl`.

## Version 0.2.13

- Improved error handling and documentation in `CountSpell`.

## Version 0.2.12

Improved error handling in `FillDown` for missing variables.

## Version 0.2.11

Allow arguments for `slide` to be passed throught `StartEnd`.

## Version 0.2.10

`slide` convert with tbl_df.

## Version 0.2.9

Fixed a bug in `slide` when using factor class variables as `GroupVar`.

## Version 0.2.8.3

Minor change to `PercChange` allowing arguments to be passed to `slide`.

Minor internal code improvements.

## Version 0.2.8.2

`FillDown` down accepts just vectors. This is useful when used with dplyr's
`group_by` and `mutate`/`summarize` functions.

## Version 0.2.8.1

Added `TimeVar` argument to `slide`. The argument is optional. When the
argument is specified, then the data is ordered by `Var`-`TimeVar` before
sliding.

- Internal code improvements.

## Version 0.2.8

Added function `TimeExpand`. This expands a data set so that it includes an
observation for each time point in a sequence. Works with grouped data.

## Version 0.2.7.2

- Error message added to `slide` when using grouped data.

## Version 0.2.7.1

Added dplyr version depedency >= 0.3.

Minor `FillIn` change checking for `missing` rather than `NULL` for `Var2`.

## Version 0.2.7

Fixed a bug in `slide` when there are invalid sliding values.

## Version 0.2.6

- Fixed a misspelling in the `CountSpell` documentation. Thanks to
Sascha Schuster.

## Version 0.2.5

- Minor error handling improvements to `grepl.sub`.

## Version 0.2.4

- Added `FillDown` a function that fills in missing (NA) values with the
previous non-missing value.

- Internal code formating improvements.

## Version 0.2.3

- `keepInvalid` argument added to `slide`. This allows the user to specify
whether or not to keep observations for groups for which no valid lag/lead can
be created due to an insufficient number of time points. If TRUE then these
groups are returned to the bottom of the data frame and NA is given for their
new lag/lead variable value.

## Version 0.2.2

Minor internal improvement where slide (and associated functions) return only
data frame class objects

## Version 0.2.1

Minor documentation improvements to `slideMA`.

Internal improvements to `MoveFront`.

## Version 0.2

Added `CountSpell`, a function that returns a variable counting the spell number
for an observation. Works with grouped data.

Added an `OnlyStart` argument to `StartStop` to return only a new `Spell_Start`
variable.

Documentation and internal code improvements.

## Version 0.1.27

`FillIn` internal code improvements.

## Version 0.1.26

!!! `grepl.sub` argument `patterns` changed to `pattern`. !!!

`InsertRow` function allows user to insert a row into a data frame.
Largely implements: http://stackoverflow.com/a/11562428.

Minor documentation improvements.

Minor internal code cleaning.

## Version 0.1.25

No longer depends on the forecast package.

## Version 0.1.24

`slide` and `slideMA` now rely on `dplyr` rather than `plyr` for speed
improvements.

## Version 0.1.23

Minor `StartEnd` bug fix.

## Version 0.1.22

`StartEnd` function added. Finds the starting and ending time points of a spell.

Small improvement to `slide` to drop temporary 'fake' variable' before returning the data frame.

## Version 0.1.21

Small improvement to `FillIn` so that it doesn't attempt to find correlations if Var2 is not numeric.

Error message improvements.

## Version 0.1.20

Added `SpreadDummy` to spread a dummy variable (1's and 0') over a specified time period and for specified groups.

Improved error handling in `slide` for when the number of rows in a group is smaller than the argument set with `slideBy`.

Minor argument default changes in `FindReplace`.

## Version 0.1.19

Added `slideMA` for creating a moving average for a period before or after each time point for a given variable.

Message improvements to `shift`

## Version 0.1.18

Added `NaVar` for creating new variable(s) indicating if there are missing values in other variable(s).

Minor internal improvements to `rmExcept`.


R/utils.R
#' Remove all objects from a workspace except those specified by the user.
#'
#' \code{rmExcept} removes all objects from a workspace except those specified
#' by the user.
#'
#' @param keepers a character vector of the names of object you would like to
#' keep in your workspace.
#' @param envir the \code{\link{environment}} to remove objects from. The
#' default is the global environment (i.e. \code{\link{globalenv}}).
#' @param message logical, whether or not to return a message informing the user
#' of which objects were removed.
#'
#' @examples
#' # Create objects
#' A <- 1; B <- 2; C <- 3
#'
#' # Remove all objects except for A
#' rmExcept("A")
#'
#' # Show workspace
#' ls()
#'
#' @export

rmExcept <- function(keepers, envir = globalenv(), message = TRUE) {
    if (class(keepers) != 'character') {
        stop('The keepers argument must be a character vector.', call. = FALSE)
    }
    DeleteObj <- setdiff(ls(envir = envir), keepers)
    rm(list = DeleteObj, envir = envir)
    if (isTRUE(message)) {
        message("Removed the following objects:\n")
        message(paste(DeleteObj, collapse = ", "))
    }
}


#' Subset a data frame if a specified pattern is found in a character string
#'
#' @param data data frame.
#' @param pattern character vector containing a regular expression to be matched
#' in the given character vector.
#' @param Var character vector of the variables that the pattern should be
#' found in.
#' @param keep.found logical. whether or not to keep observations where the
#' pattern is found (\code{TRUE}) or not found (\code{FALSE}).
#' @param ... arguments to pass to \code{\link{grepl}}.
#'
#' @examples
#' # Create data frame
#' ABData <- data.frame(a = c("London, UK", "Oxford, UK", "Berlin, DE",
#'                     "Hamburg, DE", "Oslo, NO"),
#'                     b = c(8, 0.1, 3, 2, 1))
#'
#' # Keep only data from Germany (DE)
#' ABGermany <- grepl.sub(data = ABData, pattern = "DE", Var = "a")
#'
#' @export

grepl.sub <- function(data, pattern, Var, keep.found = TRUE, ...){
    y <- NULL
    if (missing(Var)) stop('Var must be specified', call. = FALSE)

    data$y <- grepl(pattern = paste0(pattern, collapse =  "|"), x = data[, Var],
                    ...)
    subdata <- subset(data, y == keep.found)
    subdata$y <- NULL
    subdata
}

#' Replace multiple patterns found in a character string column of a data frame
#'
#' \code{FindReplace} allows you to find and replace multiple character string
#' patterns in a data frame's column.
#'
#' @param data data frame with the column you would like to replace string
#' patterns.
#' @param Var character string naming the column you would like to replace
#' string patterns. The column must be of class \code{character} or
#' \code{factor}.
#' @param replaceData a data frame with at least two columns. One contains the
#' patterns to replace and the other contains their replacement. Note: the
#' pattern and its replacement must be in the same row.
#' @param from character string naming the column with the patterns you would
#' like to replace.
#' @param to character string naming the column with the the pattern
#' replacements.
#' @param exact logical. Indicates whether to only replace exact pattern matches
#' (\code{TRUE}) or not (\code{FALSE}).
#' @param vector logical. If \code{TRUE} then the replacement is returned as a
#' single vector. If \code{FALSE} then the whole data frame is returned.
#'
#' @examples
#' # Create original data
#' ABData <- data.frame(a = c("London, UK", "Oxford, UK", "Berlin, DE",
#'                      "Hamburg, DE", "Oslo, NO"),
#'                      b = c(8, 0.1, 3, 2, 1))
#'
#' # Create replacements data frame
#' Replaces <- data.frame(from = c("UK", "DE"), to = c("England", "Germany"))
#'
#' # Replace patterns and return full data frame
#' ABNewDF <- FindReplace(data = ABData, Var = "a", replaceData = Replaces,
#'                      from = "from", to = "to", exact = FALSE)
#'
#' # Replace patterns and return the Var as a vector
#' ABNewVector <- FindReplace(data = ABData, Var = "a", replaceData = Replaces,
#'                      from = "from", to = "to", vector = TRUE)
#'
#' @export

FindReplace <- function(data, Var, replaceData, from = 'from', to = 'to',
                        exact = TRUE, vector = FALSE){
  if (!(class(data[, Var]) %in% c('character', 'factor'))) {
    stop(paste(Var, 'is not a character string or factor. Please convert to a character string or factor and then rerun.'),
         call. = FALSE)
  }
  if (isTRUE(exact)) {
    message('Only exact matches will be replaced.')
  }

  ReplaceNRows <- nrow(replaceData)

  for (i in 1:ReplaceNRows) {
    if (isTRUE(exact)) {
      data[, Var] <- gsub(pattern = paste0("^", replaceData[i, from], "$"),
                        replacement = replaceData[i, to], data[, Var])
    }
    if (!isTRUE(exact)) {
      data[, Var] <- gsub(pattern = replaceData[i, from],
                        replacement = replaceData[i, to], data[, Var])
    }
  }
  if (isTRUE(vector)) {
    data <- data[, Var]
  }
  return(data)
}

#' Drop one or more variables from a data frame.
#'
#' \code{VarDrop} drops one or more variables from a data frame.
#'
#' @param data a data frame.
#' @param Var character vector containing the names of the variables to drop.
#'
#' @examples
#' # Create dummy data
#' a <- c(1, 2, 3, 4, NA)
#' b <- c( 1, NA, 3, 4, 5)
#' c <- c(1:5)
#' d <- c(1:5)
#' ABCData <- data.frame(a, b, c, d)
#'
#' # Drop a and b
#' DroppedData <- VarDrop(ABCData, c('b', 'c'))
#'
#' @export

VarDrop <- function(data, Var) {
    if (!all(Var %in% names(data))) {
    message('At least one of the specified variables to drop is not found.')
    }
    data <- data[, !(names(data) %in% Var)]
    return(data)
}

#' Merges 2 data frames and report/drop/keeps only duplicates.
#'
#' \code{dMerge} merges 2 data frames and reports/drops/keeps only duplicates.
#'
#' @param data1 a data frame. The first data frame to merge.
#' @param data2 a data frame. The second data frame to merge.
#' @param by specifications of the columns used for merging.
#' @param Var depricated.
#' @param dropDups logical. Whether or not to drop duplicated rows based on
#' \code{Var}. If \code{dropDups = FALSE} then it gives a count of the
#' duplicated rows.
#' @param dupsOut logical. If \code{TRUE} then a data frame only containing
#' duplicated values is returned and \code{dropDups} is ignored.
#' @param fromLast logical indicating if duplication should be considered from
#' the reverse side. Only relevant if \code{dropDups = TRUE}.
#' @param all logical; all = L is shorthand for \code{all.x = L} and
#' \code{all.y = L}, where \code{L} is either \code{TRUE} or \code{FALSE}.
#' @param all.x logical; if TRUE, then extra rows will be added to the output,
#' one for each row in x that has no matching row in y. These rows will have
#' \code{NA}s in those columns that are usually filled with values from y. The
#' default is #' \code{FALSE}, so that only rows with data from both x and y are
#' included in the output.
#' @param all.y logical; analogous to \code{all.x}.
#' @param sort logical. Should the result be sorted on the by columns?
#' @param suffixes	a character vector of length 2 specifying the suffixes to be
#' used for making unique the names of columns in the result which not used for
#' merging (appearing in by etc).
#' @param incomparables	values which cannot be matched. See \code{\link{match}}.
#'
#'
#' @seealso \code{\link{duplicated}}, \code{\link{merge}}
#'
#' @export

dMerge <- function(data1, data2, by, Var, dropDups = TRUE, dupsOut = FALSE,
    fromLast = FALSE, all = FALSE, all.x = all, all.y = all,
    sort = TRUE, suffixes = c(".x",".y"),
    incomparables = NULL){
    if (isTRUE(dropDups) & isTRUE(dupsOut)) {
        message("dropDups ignored")
        dropDups <- FALSE
    }
    if (!missing(Var)) {
        warning('Var is depricated. Transfering value to "by" argument.',
                call. = FALSE)
        by <- Var
    }

    # Perform basic merge
    Comb <- merge(data1, data2, by = by, all = all, all.x = all.x,
        all.y = all.y,
        sort = TRUE, suffixes = suffixes,
        incomparables = incomparables)

    # Find duplicated
    DupDF <- Comb[duplicated(Comb[, by]), ]

    # Inform user of duplicates and drop if requested
    if (is.null(DupDF)) {
        message('There are no duplicated rows.')
        if (isTRUE(dupsOut)) {
            message('Full data set returned.')
        }
    }
    else if (!is.null(DupDF)) {
        Count <- nrow(DupDF)
        if (!isTRUE(dropDups) & !isTRUE(dupsOut)) {
            message(paste('There are', Count, 'duplicated rows.'))
        }
        else if (isTRUE(dropDups) & !isTRUE(dupsOut)) {
            Comb <- Comb[!duplicated(Comb[, Var], fromLast = fromLast), ]
            message(paste(Count, 'duplicated rows were dropped.'))
        }
    else if (!isTRUE(dropDups) & isTRUE(dupsOut)) {
        message(paste('There are', Count, 'duplicated rows.'))
        Comb <- DupDF
        }
    }
    return(Comb)
}

#' Inserts a new row into a data frame
#' @param data a data frame to insert the new row into.
#' @param NewRow a vector whose length is the same as the number of columns in
#' \code{data}.
#' @param RowNum numeric indicating which row to insert the new row as. If not
#' specified then the new row is added to the end using a vanilla
#' \code{\link{rbind}} call.
#'
#' @examples
#' # Create dummy data
#' A <- B <- C <- D <- sample(1:20, size = 20, replace = TRUE)
#' Data <- data.frame(A, B, C, D)
#'
#' # Create new row
#' New <- rep(1000, 4)
#'
#' # Insert into 4th row
#' Data <- InsertRow(Data, NewRow = New, RowNum = 4)
#'
#' @source The function largely implements:
#' \url{http://stackoverflow.com/a/11562428}
#' @export

InsertRow <- function(data, NewRow, RowNum = NULL) {
    if (ncol(data) != length(NewRow)) {
        stop('NewRow must be the same length as the number of columns in the data.\n',
             call. = FALSE)
    }
    if (!is.null(RowNum)) {
        data[seq(RowNum + 1,nrow(data) + 1), ] <- data[seq(RowNum, nrow(data)), ]
        data[RowNum, ] <- NewRow
    }
    else if (is.null(RowNum)) {
        data <- rbind(data, NewRow)
    }
    return(data)
}

#' Find duplicated values in a data frame and subset it to either include or
#' not include them.
#'
#' @param data a data frame to select the duplicated values from.
#' @param Vars character vector of variables in \code{data} to find duplicated
#' values on.
#' @param NotDups logical. If \code{TRUE} then a data frame without duplicated
#' values is returned.
#' @param test logical. If \code{TRUE} then the function will return an error
#' if there are duplicated values.
#' @param ... arguments to pass to \code{\link{duplicated}}.
#'
#' @examples
#' Data <- data.frame(ID = c(1, 1, 2, 2), Value = c(1, 2, 3, 4))
#'
#' FindDups(Data, Vars = 'ID')
#'
#' @return a data frame, unless \code{test = TRUE} and there are duplicates.
#'
#' @export

FindDups <- function(data, Vars, NotDups = FALSE, test = FALSE, ...) {
    dups <- data[duplicated(data[, Vars], ...), ]
    numb_dups <- nrow(dups)
    message(sprintf('%s duplicates in the data frame.', numb_dups))

    if (isTRUE(test) & numb_dups != 0) stop('Find your duplicates!', call. = F)

    if (!isTRUE(NotDups)) {
        if (numb_dups > 0) {
            out <- dups[, Vars]
        }
    }
    if (isTRUE(NotDups)) out <- data[!duplicated(data[, Vars], ...), ]

    if (!isTRUE(test) & numb_dups > 0) return(out)
}

#' Extract a single column from a data frame (including tbl_df) and return it as
#' a vector.
#'
#' @param data a data frame to extract the column from.
#' @param Var a character string identifying the column to extract as a vector.
#'
#' @source Modified from Tommy O'Dell:
#' \url{http://stackoverflow.com/a/24730843/1705044}
#' @return a vector
#' @export

pull <- function(data, Var){
    temp_var_name <- Var
    data[, if(is.name(substitute(temp_var_name)))
        deparse(substitute(temp_var_name))
        else temp_var_name, drop = FALSE][[1]]
}
